src/crowdin_client.cpp

```diff
@@ -135,150 +159,332 @@ CrowdinClient::~CrowdinClient() {}
 dispatch::future<void> CrowdinClient::Authenticate()
 {
     auto url = WrapLink(OAUTH_AUTHORIZE_URL);
-    m_authCallback.reset(new dispatch::promise<void>);
     wxLaunchDefaultBrowser(url);
+    lock_guard<mutex> lck(m_authMutex);
+    m_authCallback.reset(new dispatch::promise<void>);
     return m_authCallback->get_future();
 }
 
 
 void CrowdinClient::HandleOAuthCallback(const std::string& uri)
 {
-    if (!m_authCallback)
-        return;
-
-    const regex re("access_token=([^&]+)&");
+    const regex re("code=([^&]+)&state=([^&]+)");
     smatch m;
-    if (!regex_search(uri, m, re))
+    if (!regex_search(uri, m, re)
+        || m.size() < 3
+        || m.str(2) != OAUTH_STATE) {
         return;
+    }
 
-    SaveAndSetToken(m.str(1));
-
-    m_authCallback->set_value();
-    m_authCallback.reset();
+    m_oauth->post("oauth/token", json_data(json({
+        { "grant_type", "authorization_code" },
+        { "client_id", OAUTH_CLIENT_ID },
+        { "client_secret", OAUTH_CLIENT_SECRET },
+        { "redirect_uri", OAUTH_URI_PREFIX },
+        { "code", m.str(1) }
+    }))).then([this](json r) {
+        lock_guard<mutex> lck(m_authMutex);
+        if (!m_authCallback)
+           return;
+        SaveAndSetAuth(r);
+        m_authCallback->set_value();
+        m_authCallback.reset();
+    });
 }
 
+dispatch::future<void> CrowdinClient::RefreshToken()
+{
+    if(std::time(NULL) < m_authExpireTime) {
+        return dispatch::make_ready_future();
+    }
+
+    return m_oauth->post("oauth/token", json_data(json({
+        { "grant_type", "refresh_token" },
+        { "client_id", OAUTH_CLIENT_ID },
+        { "client_secret", OAUTH_CLIENT_SECRET },
+        { "refresh_token", m_authRefreshToken }
+    }))).then([this](json r) {
+        lock_guard<mutex> lck(m_authMutex);
+        SaveAndSetAuth(r);
+    });
+}
 
 bool CrowdinClient::IsOAuthCallback(const std::string& uri)
 {
     return boost::starts_with(uri, OAUTH_URI_PREFIX);
 }
 
+//TODO: validate JSON schema in all API responses
+//      and handle errors since now Poedit just crashes
+//      in case of some of expected keys missing
 
 dispatch::future<CrowdinClient::UserInfo> CrowdinClient::GetUserInfo()
 {
-    return m_api->get("/api/account/profile?json=")
+    return RefreshToken().then([=] () {
+    return m_api->get("user")
         .then([](json r)
         {
-            json profile = r["profile"];
+            cout << "\n\nGot user info: " << r << "\n\n";
```

Comment:
Sure, no problem, will remove if you prefer and finally decide to after my below proposal/argumentation.
Just how about keeping them in the code but wrapping them into something like `LOG()` macro which would just take as single argument same as passed to `cout` while enable only when  `_DEBUG` and/or `DEBUG` macro(s) defined? While on Windows and MacOS (where such GUI applications are not attached to console with stdin/out/err so far I remember so this loggings comes to nowhere) - we could direct it to e.g. to `debug.log` if above macro(s) defined.
Because it's really very useful for and essentially speeds up debugging especially at least for me :)

Reply:
If you feel logging is useful here, please use `wxLogTrace` as is done elsewhere, using `"poedit.crowdin"` as the mask.

Reply:
Done in 8b3c37c15a275a7bc299dc85d7abb9dc9a205008 but allowed myself to deviate from your way very slightly by wrapping `wxLogTrace()` to `LOG()` macro to use C++ `<<` (`ostream`) style formatting instead of C `printf` style since
* at least for me:
  * It's easy to make just trivial mechanical mistake with format specifiers to mismatch actual args about what compiler even does not warn and that is why requires to pay/strain extra attention
  * Needs explicit conversion to C string (char*) for logging `json` responses instead of just e.g. `<< response` where implicit overloaded operator is used provided by `json` lib
* I am lazy for extra typing when adding logging that is why prefer short macro defied locally like this `LOG()` :)
* less replacing just cout=>log without replacing formatting itself which is already written.

However if you feel uncomfortable with this here - please let and I will  use only `wxLogTrace()` directly

Reply:
I get that you have personal preference for the `<<` style (I've seen it elsewhere in the PR where I would use `+` to concatenate strings); I for example don't. 

My personal preference is besides the point, though: I really think that *consistency* matters for readability by future persons (that may be future-me or somebody else). If all of Poedit's codebase uses `wxLogXXXX` functions for logging and `wxLogTrace` for this kind of logging, we shouldn't do something entirely different in just one file. 

A caveat to that is that I would see it as OK to start using some other logging library with *obvious* advantages over the current setup in just one file if it were with the goal of migrating everything else to that library.

Reply:
Done exactly as you wish/expect e330ce8c8af4102c2255e1c2b30791f39fbc1499